Parameterize `balanceTransaction` over `era` type parameter

Anviking opened the pull request:
- [x] Mostly mechanical changes to parameterize `balanceTransaction` over `era`

### Motivation

- Simplifies error handling in various helper functions
- Makes it easier to in the future use a single `Cardano.UTxO era` across the entire `balanceTransaction` function.

<!-- Additional comments, links, or screenshots to attach, if any. -->

### Issue Number

ADP-1642

Anviking committed:
Don't derive Generic for TransactionLayer

When we add fields containing `forall era` in a future commit,
we cannot derive Generic.

Anviking committed:
Simplify helper functions in TransactionLayer using `era` type param

Anviking committed:
Parameterize `balanceTransaction` over `era` type param

This first commit does the minimal work to make `balanceTransaction`
take and return a `Cardano.Tx era` instead of an existential `SealedTx`.

In the next commit we will make use of the era parameter to massively
simplify helper functions in `TransactionLayer`.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3697,7 +3691,7 @@ data WalletFollowLog
 data WalletLog
     = MsgSelectionStart Int [TxOut]
-    | MsgSelectionForBalancingStart Int PartialTx
+    | MsgSelectionForBalancingStart Int String
```
👀 Would like to do something better here

Anviking replied:
Added a `BuildableInAnyEra` helper type.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -2449,24 +2451,67 @@ balanceTransaction ctx genChange (ApiT wid) body = do
+    inAnyShelleyBasedEra
+        :: Cardano.InAnyCardanoEra a
+        -> Maybe (Cardano.InAnyShelleyBasedEra a)
+    inAnyShelleyBasedEra = \case
```
Also used in tests; could be re-used better

Anviking replied:
Now done.

Anviking commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -126,6 +125,7 @@ import Fmt
     ( Generic )
 
+import qualified Cardano.Api as Cardano
 import qualified Cardano.Api.Shelley as Node
```
Could drop, and replace `Node` with `Cardano` for consistency

Anviking replied:
Done.

Anviking commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1120,29 +1088,30 @@ type AlonzoTx =
     -> [Redeemer]
-    -> SealedTx
-    -> Either ErrAssignRedeemers SealedTx
+    -> (Cardano.Tx era)
```
```suggestion
    -> Cardano.Tx era
```

Anviking committed:
Rename import in Cardano.Wallet.Transaction module

Anviking committed:
Add Cardano.Api.Extra module

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3435,11 +3429,12 @@ data ErrBalanceTx
 
 -- TODO: Remove once problems are fixed.
 data BalanceTxNotSupportedReason
-    = UnderestimatedFee Coin SealedTx
```
👀 Not being able to print this as a counterexample in properties would be a regression

Anviking replied:
Fixed. I kept the `SealedTx`.

Anviking commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2809,33 +2809,28 @@ instance Arbitrary PartialTx where
             inputs
             redeemers
-        | tx' <- shrinkTx (alonzoCardanoTx tx)
+        | tx' <- shrinkTx tx -- TODO: Need to take tail to fix shrinker?
```
👀

Anviking replied:
Seems to be correct because we never use more than one `<-` in each list comprehension.

Anviking commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -4821,12 +4842,6 @@ instance IsServerError ErrUpdateSealedTx where
                 , "Please make sure to remove all key witnesses from the request."
                 ]
-        ErrByronTxNotSupported{} ->
-            apiError err501 NotImplemented  $ T.unwords
```
We used to have two different ways to render "byron not supported" 🙀

I think err501 is wrong:
https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/501

However, perhaps 405 should be used instead of 403 on line 4324...

> If the server does recognize the method, but intentionally does not support it, the appropriate response is [405 Method Not Allowed](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405).

Anviking replied:
Will consider changing tomorrow.

Anviking replied:
On second look, err400 doesn't seem right either. I'm sticking with err403, which we already are using a lot.

jonathanknowles commented on lib/core/src/Cardano/Api/Extra.hs:
```diff
@@ -0,0 +1,55 @@
+    , Tx
+    )
+
+-- | Apply an era-parameterized function to an existentially wrapped
```
```suggestion
-- | Apply an era-parameterized function to an existentially-wrapped
```

jonathanknowles committed:
Move function `modifyLedgerTx` to the top-level.

This function is already extremely nested, making it hard to read.

jonathanknowles committed:
Move function `adjustBody` to top-level and rename to `adjustBodyShelley`.

jonathanknowles committed:
Avoid record deconstruction in function argument list.

Deconstructing a complex record within the definition of a function
makes the arguments quite hard to read, as when new fields are added
the line starts extending beyong the edge of the line limit.

Instead, deconstruct these values in a `where` or an `let` clause.

jonathanknowles committed:
Collapse some unnecessary double indents.

jonathanknowles committed:
Wrap long lines, and unwrap lines that no longer need wrapping.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1720,56 +1724,50 @@ balanceTransactionWithSelectionStrategy
         else throwE $ ErrBalanceTxInternalError $ ErrFailedBalancing bal
 
-    txBalance :: SealedTx -> ExceptT ErrBalanceTx m Cardano.Value
+    txBalance :: Cardano.Tx era -> Cardano.Value
```
Nice! 👍🏻

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -1720,56 +1724,50 @@ balanceTransactionWithSelectionStrategy
-        case evaluateTransactionBalance tl tx nodePParams utxo externalInputs of
-            Just x -> pure x
-            Nothing -> throwE $ ErrBalanceTxUpdateError ErrByronTxNotSupported
+        evaluateTransactionBalance tl tx nodePParams utxo externalInputs
```
🎉

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3700,10 +3698,29 @@ data WalletFollowLog
+    , Buildable (tx era)
+    ) => BuildableInAnyEra (Cardano.CardanoEra era) (tx era)
+
+instance Show (BuildableInAnyEra a) where
+    show (BuildableInAnyEra _ a) = show a
```
I think it's possible to derive this instance automatically, if you add the `StandaloneDeriving` extension:
```suggestion
deriving instance Show (BuildableInAnyEra a)
```

Anviking replied:
Sticking with the original to avoid the wrapper constructor cluttering the logs.

Anviking commented on lib/core/src/Cardano/Api/Extra.hs:
```diff
@@ -0,0 +1,55 @@
+    = f tx
+
+-- | "Upcast" an existentially wrapped tx.
+asAnyEraTx :: forall era. IsShelleyBasedEra era => Tx era -> InAnyCardanoEra Tx
```
Hm. Might want to tweak the wording. This is not quite the inverse of `asAnyShelleyBasedEra` because this function takes a `forall era. Tx era` rather than a `InAnyShelleyBasedEra`.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -3436,6 +3433,7 @@ data ErrBalanceTx
     | ErrBalanceTxAssignRedeemers ErrAssignRedeemers
     | ErrBalanceTxInternalError ErrBalanceTxInternalError
     | ErrBalanceTxZeroAdaOutput
+    | ErrByronTxNotSupported
```
Perhaps this could be `ErrTxEraNotSupported`, with a parameter to express the era?

Mentioning Byron here seems a bit "specific" for the generality of the `ErrBalanceTx` type.

jonathanknowles replied:
Perhaps you could have:

```suggestion
    | ErrEraNotSupported EraNotSupported

data EraNotSupported = ErrByronTxNotSupported
```

Anviking replied:
I really _do_ like avoiding unnecessary nesting. Can revisit in the future if needed.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Transaction.hs:
```diff
@@ -509,11 +507,10 @@ data ErrCannotQuit
     = ErrExistingKeyWitnesses Int
     -- ^ The `SealedTx` couldn't not be updated because the *n* existing
     -- key-witnesses would have been rendered invalid.
-    | ErrByronTxNotSupported
```
👍🏻

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1259,7 +1229,7 @@ _assignScriptRedeemers (toAlonzoPParams -> pparams) ti resolveInput redeemers tx
     -- final redeemers, if any.
     addScriptIntegrityHash
-        :: forall era. (era ~ Cardano.ShelleyLedgerEra Cardano.AlonzoEra)
+        :: forall e. (e ~ Cardano.ShelleyLedgerEra Cardano.AlonzoEra)
```
I'm curious as to why you changed this. Does it clash with another `era` definition?

Anviking replied:
Yes

jonathanknowles commented on lib/core/src/Cardano/Api/Extra.hs:
```diff
@@ -0,0 +1,55 @@
+
+-- | "Upcast" an existentially wrapped tx.
+asAnyEraTx :: forall era. IsShelleyBasedEra era => Tx era -> InAnyCardanoEra Tx
+asAnyEraTx = InAnyCardanoEra cardanoEra
```
```suggestion
inAnyCardanoEra :: IsCardanoEra era => Tx era -> InAnyCardanoEra Tx
inAnyCardanoEra = InAnyCardanoEra cardanoEra
```

jonathanknowles committed:
Revise names of `updateSealedTx` helper functions.

The name `updateSealedTx` is itself an anachronism, because it no longer
acts on a `SealedTx`. However, updating that now would require changes
in many files. So this commit leaves it alone.
